File: Souvik/MLP.py
import numpy as np

class MLP:
    def __init__(self, input_size, hidden_size, output_size):
        self.input_size = input_size
        self.hidden_size = hidden_size
        self.output_size = output_size

        # Initialize weights and biases
        self.weights_input_hidden = np.random.randn(self.input_size, self.hidden_size)
        self.bias_hidden = np.random.randn(self.hidden_size)
        self.weights_hidden_output = np.random.randn(self.hidden_size, self.output_size)
        self.bias_output = np.random.randn(self.output_size)

    def sigmoid(self, x):
        return 1 / (1 + np.exp(-x))

    def sigmoid_derivative(self, x):
        return x * (1 - x)

    def forward(self, X):
        # Input to Hidden Layer
        self.hidden_input = np.dot(X, self.weights_input_hidden) + self.bias_hidden
        self.hidden_output = self.sigmoid(self.hidden_input)

        # Hidden to Output Layer
        self.output_input = np.dot(self.hidden_output, self.weights_hidden_output) + self.bias_output
        self.output = self.sigmoid(self.output_input)

        return self.output

    def backward(self, X, y, output, learning_rate):
        # Calculate output layer error
        output_error = y - output
        output_delta = output_error * self.sigmoid_derivative(output)

        # Calculate hidden layer error
        hidden_error = output_delta.dot(self.weights_hidden_output.T)
        hidden_delta = hidden_error * self.sigmoid_derivative(self.hidden_output)

        # Update weights and biases
        self.weights_hidden_output += self.hidden_output.T.dot(output_delta) * learning_rate
        self.bias_output += np.sum(output_delta, axis=0) * learning_rate
        self.weights_input_hidden += X.T.dot(hidden_delta) * learning_rate
        self.bias_hidden += np.sum(hidden_delta, axis=0) * learning_rate

    def train(self, X, y, epochs, learning_rate):
        for epoch in range(epochs):
            output = self.forward(X)
            self.backward(X, y, output, learning_rate)

    def predict(self, X):
        return self.forward(X)

# Example usage:
if __name__ == "__main__":
    # XOR Problem
    X = np.array([[0, 0], [0, 1], [1, 0], [1, 1]])
    y = np.array([[0], [1], [1], [0]])

    mlp = MLP(input_size=2, hidden_size=2, output_size=1)
    mlp.train(X, y, epochs=10000, learning_rate=0.1)

    predictions = mlp.predict(X)
    print("Predictions:")
    print(predictions)

File: Souvik/MLP.java
import java.util.Arrays;
import java.util.Random;

public class MLP {

    private int inputSize;
    private int hiddenSize;
    private int outputSize;
    private double[][] weightsInputHidden;
    private double[] biasHidden;
    private double[][] weightsHiddenOutput;
    private double[] biasOutput;
    private double[] hiddenLayerOutput;
    private double[] outputLayerOutput;
    private Random random;

    public MLP(int inputSize, int hiddenSize, int outputSize) {
        this.inputSize = inputSize;
        this.hiddenSize = hiddenSize;
        this.outputSize = outputSize;
        this.random = new Random();

        // Initialize weights and biases
        weightsInputHidden = new double[inputSize][hiddenSize];
        biasHidden = new double[hiddenSize];
        weightsHiddenOutput = new double[hiddenSize][outputSize];
        biasOutput = new double[outputSize];
        hiddenLayerOutput = new double[hiddenSize];
        outputLayerOutput = new double[outputSize];

        initializeWeights(weightsInputHidden);
        initializeWeights(weightsHiddenOutput);
        initializeBias(biasHidden);
        initializeBias(biasOutput);
    }

    private void initializeWeights(double[][] weights) {
        for (int i = 0; i < weights.length; i++) {
            for (int j = 0; j < weights[i].length; j++) {
                weights[i][j] = random.nextDouble() - 0.5;
            }
        }
    }

    private void initializeBias(double[] bias) {
        for (int i = 0; i < bias.length; i++) {
            bias[i] = random.nextDouble() - 0.5;
        }
    }

    private double sigmoid(double x) {
        return 1 / (1 + Math.exp(-x));
    }

    private double sigmoidDerivative(double x) {
        return x * (1 - x);
    }

    public double[] forward(double[] input) {
        // Input to Hidden Layer
        for (int j = 0; j < hiddenSize; j++) {
            double sum = 0;
            for (int i = 0; i < inputSize; i++) {
                sum += input[i] * weightsInputHidden[i][j];
            }
            sum += biasHidden[j];
            hiddenLayerOutput[j] = sigmoid(sum);
        }

        // Hidden to Output Layer
        for (int k = 0; k < outputSize; k++) {
            double sum = 0;
            for (int j = 0; j < hiddenSize; j++) {
                sum += hiddenLayerOutput[j] * weightsHiddenOutput[j][k];
            }
            sum += biasOutput[k];
            outputLayerOutput[k] = sigmoid(sum);
        }

        return outputLayerOutput;
    }

    public void backward(double[] input, double[] target, double[] output, double learningRate) {
        double[] outputErrors = new double[outputSize];
        double[] outputDeltas = new double[outputSize];

        for (int k = 0; k < outputSize; k++) {
            outputErrors[k] = target[k] - output[k];
            outputDeltas[k] = outputErrors[k] * sigmoidDerivative(output[k]);
        }

        double[] hiddenErrors = new double[hiddenSize];
        double[] hiddenDeltas = new double[hiddenSize];

        for (int j = 0; j < hiddenSize; j++) {
            double sum = 0;
            for (int k = 0; k < outputSize; k++) {
                sum += outputDeltas[k] * weightsHiddenOutput[j][k];
            }
            hiddenErrors[j] = sum;
            hiddenDeltas[j] = hiddenErrors[j] * sigmoidDerivative(hiddenLayerOutput[j]);
        }

        for (int j = 0; j < hiddenSize; j++) {
            for (int k = 0; k < outputSize; k++) {
                weightsHiddenOutput[j][k] += learningRate * outputDeltas[k] * hiddenLayerOutput[j];
            }
        }

        for (int k = 0; k < outputSize; k++) {
            biasOutput[k] += learningRate * outputDeltas[k];
        }

        for (int i = 0; i < inputSize; i++) {
            for (int j = 0; j < hiddenSize; j++) {
                weightsInputHidden[i][j] += learningRate * hiddenDeltas[j] * input[i];
            }
        }

        for (int j = 0; j < hiddenSize; j++) {
            biasHidden[j] += learningRate * hiddenDeltas[j];
        }
    }

    public void train(double[][] inputs, double[][] targets, int epochs, double learningRate) {
        for (int epoch = 0; epoch < epochs; epoch++) {
            for (int i = 0; i < inputs.length; i++) {
                double[] output = forward(inputs[i]);
                backward(inputs[i], targets[i], output, learningRate);
            }
        }
    }

    public double[] predict(double[] input) {
        return forward(input);
    }

    public static void main(String[] args) {
        // XOR Problem
        double[][] inputs = {
            {0, 0},
            {0, 1},
            {1, 0},
            {1, 1}
        };
        double[][] targets = {
            {0},
            {1},
            {1},
            {0}
        };

        MLP mlp = new MLP(2, 2, 1);
        mlp.train(inputs, targets, 10000, 0.1);

        for (double[] input : inputs) {
            double[] output = mlp.predict(input);
            System.out.println(Arrays.toString(output));
        }
    }
}